src/valgrind_selftest/valgrind_selftest.py

```diff
@@ -0,0 +1,143 @@
+
+    def run(self, exe_path: str):
+        """ Run the test and return whether it succeeded """
+
+        result = run_with_valgrind([exe_path, self.name])
+        self.stdout = result.stdout
+        self.stderr = result.stderr
+
+        if result.returncode == 1:
+            # The valgrind_selftest program reported an error
+            self.status = self.Status.WARNING
+        else:
+            failed = result.returncode != 0
+            if failed != self.expect_failure:
+                # Valgrind reported an error (or not), but we expected the opposite
+                self.status = self.Status.ERROR
+            else:
+                # Everything behaved as expected
+                self.status = self.Status.OK
+
+        return self.status
+
+    @staticmethod
+    def read_test_list(valgrind_selftest_test_list: str) -> list[Self]:
+        """ Read the list of tests from the output of `valgrind_selftest --list`. """
+
+        return [ValgrindTest.from_line(line) for line in valgrind_selftest_test_list.split("\n")[2:] if line]
+
+
+def main(): # pylint: disable=missing-function-docstring
+    parser = argparse.ArgumentParser("valgrind_selftests")
+    parser.add_argument("valgrind_selftest_path", help="Path to the valgrind_selftest executable")
+    parser.add_argument("--filter", help="Regex filter for test names", default=".*")
+    parser.add_argument("--build-config-path", help="Path to Botan's build-config.json file", default="")
+
+    args = parser.parse_args()
+
+    # Check if the path is valid
+    if not os.path.isfile(args.valgrind_selftest_path):
+        raise FileNotFoundError(f"Invalid path: {args.valgrind_selftest_path}")
+
+    def find_test_list():
+        test_list_args = [args.valgrind_selftest_path, "--list"]
+
+        test_list_result = run_command(test_list_args)
+        if test_list_result.returncode != 0:
+            raise RuntimeError("Failed to collect the test list from valgrind_selftest")
+        test_list = ValgrindTest.read_test_list(test_list_result.stdout)
+
+        if args.build_config_path:
+            if not os.path.isfile(args.build_config_path):
+                raise FileNotFoundError(f"Invalid path: {args.build_config_path}")
+
+            # Figure out what build configuration we're running under, to decide
+            # whether or not to run certain special-case valgrind tests.
+            with open(args.build_config_path, encoding="utf-8") as build_info_file:
+                build_info = json.load(build_info_file)
+                cc_compile_flags = build_info["cc_compile_flags"]
+                cc_macro = build_info["cc_macro"]
+
+                # This is certainly not an exhaustive list of compiler configurations
+                # that could be problematic for this specific test
+                if cc_macro == "CLANG" and "-Os" in cc_compile_flags:
+                    test_list.append(ValgrindTest("clang_vs_bare_metal_ct_mask", True))
```

Comment:
As a sanity check, we could run `./botan_valgrind_selftest --list-special` and double-check that the "clang_vs_bare_metal_ct_mask" test is in this list. Also, we could take the info whether or not it is _expected to fail_ from there, rather than hard-coding it.